<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using triad.expression • triad.expression</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using triad.expression">
<meta property="og:description" content="triad.expression">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">triad.expression</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/triad-balance.html">triad-balance.Rmd</a>
    </li>
    <li>
      <a href="../articles/user-guide.html">Using triad.expression</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="user-guide_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using triad.expression</h1>
                        <h4 class="author">Ben J. Ward</h4>
            
            <h4 class="date">2021-10-07</h4>
      
      
      <div class="hidden name"><code>user-guide.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>triad.expression is an R package used to do expression analyses of so called Triads.</p>
<p>The wheat genome has three sub-genomes called A, B, and D.</p>
<p>Triads are sets of three genes, where there is a homolog in each of the three sub-genomes.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>triad.expression is not published on CRAN so you will have to install triad.expression yourself from this GitHub repository. But that’s ok, it’s easy enough.</p>
<p>The easiest way to do this is to use the <code>install_github</code> command from the R devtools package like so:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>If you do not have devtools installed, no problem, that is available through CRAN and so using</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span> </code></pre></div>
<p>in an R session should install it for you.</p>
<p>Once installed, use the package with <code>library</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">triad.expression</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="co">#&gt; Registered S3 methods overwritten by 'ggtern':</span>
<span class="co">#&gt;   method           from   </span>
<span class="co">#&gt;   grid.draw.ggplot ggplot2</span>
<span class="co">#&gt;   plot.ggplot      ggplot2</span>
<span class="co">#&gt;   print.ggplot     ggplot2</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span> <span class="co"># dplyr is going to be useful too.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span></code></pre></div>
</div>
<div id="included-data" class="section level1">
<h1 class="hasAnchor">
<a href="#included-data" class="anchor"></a>Included data</h1>
<p>The package has some data files included that have been used in our research. This package includes both processed tidy versions, as well as raw files.</p>
<div id="triad-homology" class="section level2">
<h2 class="hasAnchor">
<a href="#triad-homology" class="anchor"></a>Triad homology</h2>
<p>Triads are made up of three homologous genes from the A, B, and D sub-genomes of wheat.</p>
<p>This package makes use of, and includes, a dataset that describes these homologies.</p>
<p>To load it for use, simply use <code><a href="https://rdrr.io/r/utils/data.html">data(triad_homology)</a></code> when the package is loaded, to put it into your environment.</p>
<p>The raw version of this data is in <strong>“inst/extdata/170521_pantrans_commonref_homology.csv”</strong>.</p>
</div>
<div id="gene-locations" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-locations" class="anchor"></a>Gene locations</h2>
<p>It is important for some analyses, to know where the genes for each triad are located in the genome.</p>
<p>This data is included in this package for several varieties of wheat.</p>
<p>This data does not include any “chromosome unknown” triads from the annnotation process.</p>
<p>There is one file for each wheat variety.</p>
<p>To load it for use, simply use <code><a href="https://rdrr.io/r/utils/data.html">data(gene_locations)</a></code> when the package is loaded, to put it into your environment.</p>
<p>The raw versions of these tables are kept in <strong>“inst/extdata/triad_locations/”</strong></p>
</div>
<div id="gene-expression-data" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-expression-data" class="anchor"></a>Gene expression data</h2>
<p>Analyses in the package make use of gene expression data for the triads.</p>
<p>This data includes both the raw expression values obtained from experiment, as well as a table of metadata about the samples and the experiments they came from.</p>
<p>To get the metadata simply use <code><a href="https://rdrr.io/r/utils/data.html">data(expression_metadata)</a></code> when the package is loaded, to put it into your environment.</p>
</div>
</div>
<div id="analysing-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#analysing-expression" class="anchor"></a>Analysing expression</h1>
<div id="computing-mean-gene-expression-for-triads" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-mean-gene-expression-for-triads" class="anchor"></a>Computing mean gene expression for triads</h2>
<p>3 datasets are required, the gene expression data and its metadata, and the triad homology information.</p>
<p>All three of these tables are described in the “Included and required data” section.</p>
<p>We will use the ones provided with the package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">?</span><span class="va">expression_data</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">expression_metadata</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">triad_homology</span><span class="op">)</span></code></pre></div>
<p>Next we can compute the mean expression values for each triad gene, for every level of a factor, we use the function called <code>triad_expression_means_by_factors</code>.</p>
<p>This function takes the expression datasets, and the triad homology information, and a list of metadata column names to use as factor groupings.</p>
<p>For example, in <code>expression_metadata</code>, there is a column called “High.level.variety”, with a level for each variety of wheat.</p>
<p>To compute the mean expression level for each triad gene, for each variety then, we can use the function like so, passing in the expression and homology data, and “High.level.variety” as our factor.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meanExpressionByVariety</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/triad_expression_means_by_factors.html">triad_expression_means_by_factors</a></span><span class="op">(</span><span class="va">expression_data</span>, <span class="va">expression_metadata</span>, <span class="va">triad_homology</span>, <span class="st">"High.level.variety"</span><span class="op">)</span></code></pre></div>
<p>But what if, for example, you had multiple factors? Say you wanted to look at triad expression between varieties, but also for different tissues?</p>
<p>The metadata table has a column called “High.level.tissue”, and so clearly some expression experiments used samples from different wheat tissues. The column has several levels such as “roots”, “leaves/shoots”, etc.</p>
<p>No problem! The function we just used, can handle multiple factors:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meanExpressionByVarietyAndTissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/triad_expression_means_by_factors.html">triad_expression_means_by_factors</a></span><span class="op">(</span><span class="va">expression_data</span>, <span class="va">expression_metadata</span>, <span class="va">triad_homology</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"High.level.variety"</span>, <span class="st">"High.level.tissue"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The resulting table contains the mean expression of each gene, for every combination of factor levels - in this case “High.level.variety” and “High.level.tissue”. Now we can do some other things with it.</p>
</div>
<div id="calculating-triad-expression-balance" class="section level2">
<h2 class="hasAnchor">
<a href="#calculating-triad-expression-balance" class="anchor"></a>Calculating Triad expression balance</h2>
<p>Next we have to normalize the data, which is done with <code>normalize_triad_expression_means</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">normalizedMeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_triad_expression_means.html">normalize_triad_expression_means</a></span><span class="op">(</span><span class="va">meanExpressionByVarietyAndTissue</span><span class="op">)</span></code></pre></div>
<p>By default, this function is able to work out which columns are factors from the previous step (in this instance, “High.level.tissue” and “High.level.variety”) but you can optionally override this should it be necessary (…it probably won’t be).</p>
<p>Once you have normalized mean triad expression data, you can compute each triads distance from the centroids, and from the centroids, assign to each triad, one of 7 expression categories:</p>
<ul>
<li>Central</li>
<li>A dominant</li>
<li>B dominant</li>
<li>C dominant</li>
<li>A suppressed</li>
<li>B suppressed</li>
<li>D suppressed</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cDistances</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/triad_balance.html">triad_balance</a></span><span class="op">(</span><span class="va">normalizedMeans</span><span class="op">)</span></code></pre></div>
<p>By default, this function is able to work out which columns are factors from the previous step (in this instance, “High.level.tissue” and “High.level.variety”) but you can optionally override this should it be necessary (…it probably won’t be).</p>
</div>
<div id="plotting-triad-expression-balance-in-varieties-across-chromosomes" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-triad-expression-balance-in-varieties-across-chromosomes" class="anchor"></a>Plotting triad expression balance in varieties across chromosomes</h2>
<p>Ok, having gone through how to calculate triad expression statistics, let’s start from the beginning, and compute triad mean expression across varieties, and compute the expression categories from the centroid distances.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">expression_metadata</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">triad_homology</span><span class="op">)</span>

<span class="va">meanExpressionByVariety</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/triad_expression_means_by_factors.html">triad_expression_means_by_factors</a></span><span class="op">(</span><span class="va">expression_data</span>, <span class="va">expression_metadata</span>, <span class="va">triad_homology</span>, <span class="st">"High.level.variety"</span><span class="op">)</span>

<span class="va">normalizedMeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize_triad_expression_means.html">normalize_triad_expression_means</a></span><span class="op">(</span><span class="va">meanExpressionByVariety</span><span class="op">)</span>

<span class="va">cDistances</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/triad_balance.html">triad_balance</a></span><span class="op">(</span><span class="va">normalizedMeans</span><span class="op">)</span></code></pre></div>
<p>Next, we join the chromosomal location information to these distances.</p>
<p>So first we get the genome annotation data included in the package.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">gene_locations</span><span class="op">)</span>
<span class="va">distWithLocation</span> <span class="op">&lt;-</span> <span class="fu">join_distances_and_annotation</span><span class="op">(</span><span class="va">cDistances</span>, <span class="va">gene_locations</span><span class="op">)</span></code></pre></div>
<p>Alright, now we have the information, we can do some plotting. Most of this is handled by the <code>loom_plot</code> function, which, depending on how you preprocess some of the data, can allow you to show various things along chromosomes.</p>
<p>We’re going to use it to show the expression differences between triad genes of different varieties.</p>
<p>So first we pick a chromosome and make sure the data only represents that chromosome. This is the first filter statement.</p>
<p>Next we only really want to see genes from triads where there is in fact a difference in expression between varieties. That’s what the <code>group_by</code> and additional filter line are about.</p>
<p>The loom plot accepts the data as the first argument (here provided by the pipe). It then accepts a name of the column to use as the chromosome co-ordinates (in this case “start”), the column to use to plot the y axis (in this case “High.level.variety”), the column to use to connect points (in this case “group_id” - which indicates triad ID), and a column to use to provide colour, in this case the expression type (“clust.description”). We also provide an optional label to use as the Y axis label.</p>
<p>Note this figure may be small or squished to fit the vignette in a small amount of space.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">distWithLocation</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"chr2A"</span><span class="op">)</span>  <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">group_id</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span><span class="op">(</span><span class="va">clust.description</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">loom_plot</span><span class="op">(</span><span class="st">"start"</span>, <span class="st">"High.level.variety"</span>, <span class="st">"group_id"</span>, <span class="st">"clust.description"</span>, yLab <span class="op">=</span> <span class="st">"Variety"</span><span class="op">)</span></code></pre></div>
<p><img src="user-guide_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>Here’s a slightly more involved usage of filter, where I restrict data to only show cases where there is a clear suppression of a triad (either in A, B, or D) in some varieties. I do this by filtering the data to triads where both the phrase “suppressed” and “dominant” appear in the “clust.description” column.</p>
<p>Note this figure may be small or squished to fit the vignette in a small amount of space.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">distWithLocation</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"chr2A"</span><span class="op">)</span>  <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">group_id</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span>
    <span class="st">"suppressed"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">clust.description</span><span class="op">)</span>, <span class="fl">3</span>, <span class="fl">100L</span><span class="op">)</span>,
    <span class="st">"dominant"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">clust.description</span><span class="op">)</span>, <span class="fl">3</span>, <span class="fl">100L</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">loom_plot</span><span class="op">(</span><span class="st">"start"</span>, <span class="st">"High.level.variety"</span>, <span class="st">"group_id"</span>, <span class="st">"clust.description"</span>, yLab <span class="op">=</span> <span class="st">"Variety"</span><span class="op">)</span></code></pre></div>
<p><img src="user-guide_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
</div>
<div id="plotting-features-across-a-b-and-d-subgenomes" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-features-across-a-b-and-d-subgenomes" class="anchor"></a>Plotting features across A, B, and D subgenomes</h2>
<p>In this example we will be plotting circadian rhythm data for triads across the</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ben Ward.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
